enhancements/dns/nodeservice-etc-hosts-injection.md

```diff
@@ -0,0 +1,422 @@
+    // on all nodes. This is useful for services that need to be resolvable
+    // before DNS is fully operational, such as during cluster bootstrap or
+    // upgrades.
+    //
+    // Each service's ClusterIP will be added to /etc/hosts with an entry
+    // in the format: "<ClusterIP> <service-name>.<namespace>.svc"
+    //
+    // If not specified, a default list of critical services will be used
+    // to maintain backward compatibility.
+    //
+    // +optional
+    // +listType=map
+    // +listMapKey=namespace
+    // +listMapKey=name
+    NodeServices []NodeService `json:"nodeServices,omitempty"`
+}
+```
+
+### User Stories
+
+#### Story 1: Custom Image Registry Resolution
+
+As a cluster administrator managing a custom internal image registry service, I want to ensure that the container runtime on all nodes can resolve my registry's service name even during cluster upgrades when DNS may be temporarily unavailable, so that image pulls for core components can succeed reliably.
+
+Example:
+```yaml
+apiVersion: operator.openshift.io/v1
+kind: DNS
+metadata:
+  name: default
+spec:
+  nodeServices:
+  - namespace: my-custom-namespace
+    name: my-custom-registry
+```
+
+#### Story 2: Cluster provided infrastructure services
+
+As a platform team managing critical infrastructure services that need to be available to node scoped processes we want /etc/hosts entries to be populated to provide these entries 
+
+Example:
+```yaml
+apiVersion: operator.openshift.io/v1
+kind: DNS
+metadata:
+  name: default
+spec:
+  nodeServices:
+  - namespace: coredump
+    name: datastore
+```
+
+### Implementation Details/Notes/Constraints
+
+#### DNS Node Resolver Implementation
+
+The DNS node resolver (running as a DaemonSet on all nodes) will:
+
+1. The cluster-dns-operator watches the `DNS` custom resource for changes to the `nodeServices` field
+2. For each service specified in `nodeServices`, the operator constructs a list of service names, the internal
+   registry service `image-registry.openshift-image-registry.svc` is always added to the list
+   - Format: Service names in `service-name.namespace.svc` format
+   - Example: `datastore.coredump.svc`
+3. The operator populates the node-resolver DaemonSet `SERVICES` environment variable with the list of service names
+   - Format: Comma-separated and/or space-separated list of service names in `name.namespace.svc` format
+   - Example: `SERVICES="image-registry.openshift-image-registry.svc, datastore.coredump.svc"`
+   - The script appends `.${CLUSTER_DOMAIN}` when performing DNS lookups to create the full FQDN
+4. The DNS node resolver pods read the `SERVICES` environment variable on startup
+5. The `update-node-resolver.sh` script runs on each node and:
+   - Parses the list of service names from the `SERVICES` environment variable (comma/space-separated)
+   - For each service, performs DNS lookups using `dig` to resolve the ClusterIP
```

Comment:
Given this is about providing an option prior to DNS services being completely operational, does this work? What DNS is going to be available at this point?